pkg/operator/clustermembercontroller2/clustermembercontroller.go

```diff
@@ -0,0 +1,382 @@
+	c.queue.AddRateLimited(dsKey)
+
+	return true
+}
+
+func (c *ClusterMemberController) sync() error {
+	err := c.reconcileMembers()
+	if err != nil {
+		_, _, updateErr := v1helpers.UpdateStatus(c.operatorClient, v1helpers.UpdateConditionFn(operatorv1.OperatorCondition{
+			Type:    "ClusterMemberController2Degraded",
+			Status:  operatorv1.ConditionTrue,
+			Reason:  "Error",
+			Message: err.Error(),
+		}))
+		if updateErr != nil {
+			c.eventRecorder.Warning("ClusterMemberController2UpdatingStatus", updateErr.Error())
+		}
+		return err
+	}
+
+	_, _, updateErr := v1helpers.UpdateStatus(c.operatorClient,
+		v1helpers.UpdateConditionFn(operatorv1.OperatorCondition{
+			Type:   "ClusterMemberController2Degraded",
+			Status: operatorv1.ConditionFalse,
+			Reason: "AsExpected",
+		}))
+	return updateErr
+}
+
+func (c *ClusterMemberController) reconcileMembers() error {
+	etcdHealthy, err := c.isEtcdHealthy()
+	if err != nil {
+		return err
+	}
+	if !etcdHealthy {
+		c.eventRecorder.Eventf("WaitingOnEtcdMember", "waiting for all member of etcd to be healthy")
+		return nil
+	}
+
+	// etcd is healthy, decide if we need to scale
+	peerFQDN, dnsErrs := c.getPeerToScale()
+	if len(dnsErrs) != 0 {
+		_, _, updateErr := v1helpers.UpdateStatus(c.operatorClient, v1helpers.UpdateConditionFn(operatorv1.OperatorCondition{
+			Type:    "ClusterMemberControllerDNSLookupDegraded",
+			Status:  operatorv1.ConditionTrue,
+			Reason:  "ErrorGettingDNSNames",
+			Message: fmt.Sprintf("errod resolving etcd dns names: %#v", dnsErrs),
+		}))
+		if updateErr != nil {
+			c.eventRecorder.Warning("ClusterMemberController2UpdatingStatus", updateErr.Error())
+			return updateErr
+		}
+	} else {
+		_, _, updateErr := v1helpers.UpdateStatus(c.operatorClient, v1helpers.UpdateConditionFn(operatorv1.OperatorCondition{
+			Type:   "ClusterMemberControllerDNSLookupDegraded",
+			Status: operatorv1.ConditionFalse,
+			Reason: "AsExpected",
+		}))
+		if updateErr != nil {
+			c.eventRecorder.Warning("ClusterMemberController2UpdatingStatus", updateErr.Error())
+			return updateErr
+		}
+	}
+	if peerFQDN == "" && c.numberOfReadyMembers == totalDesiredEtcd {
+		_, _, updateErr := v1helpers.UpdateStatus(c.operatorClient,
+			v1helpers.UpdateConditionFn(operatorv1.OperatorCondition{
+				Type:    "ClusterMemberControllerScalingProgressing",
+				Status:  operatorv1.ConditionFalse,
+				Reason:  "AsExpected",
+				Message: "Scaling etcd membership completed",
+			}),
+			// todo: remove this make bootsrap remove independent
+			v1helpers.UpdateConditionFn(operatorv1.OperatorCondition{
+				Type:    "BootstrapSafeToRemove",
+				Status:  operatorv1.ConditionTrue,
+				Reason:  "AsExpected",
+				Message: "Scaling etcd membership has completed",
+			}))
+		if updateErr != nil {
+			c.eventRecorder.Warning("ClusterMemberController2UpdatingStatus", updateErr.Error())
+			return updateErr
+		}
+		// no more work left to do
+		return nil
+	}
+
```

Comment:
I don't see where you're getting the list of current etcd members to diff against the unready list.

Reply:
@deads2k I am not diffing current etcd healthy members with unready pod list because they will be disjoint sets right? I assumed that if all the existing etcd member are healthy [here](https://github.com/openshift/cluster-etcd-operator/pull/108/files#diff-5abd388ac5e01403a5c07812defe75d4R156), its corresponding pods will be in a ready state, is that an invalid assumption to make?